### PR TITLE
[Azure MySQL] Show properties feature - phrase II

### PR DESCRIPTION
**Pull request by shenqianjin**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        connectionStringsJDBC.getOutputTextArea().setText(getConnectionString(MYSQL_OUTPUT_TEXT_PATTERN_JDBC, null, null, null));
+        connectionStringsSpring.getTitleLabel().setText("Spring");
+        connectionStringsSpring.getOutputTextArea().setText(getConnectionString(MYSQL_OUTPUT_TEXT_PATTERN_SPRING, null, null, null));
+        init();
+        initListeners();
+    }
+
+    private String getConnectionString(final String pattern, final String hostname, final String database, final String username) {
+        if (StringUtils.isBlank(hostname) || StringUtils.isBlank(database) || StringUtils.isBlank(username)) {
+            return String.format(pattern, "{your_hostname}", "{your_database}", "{your_username}");
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:100]: create a default connection string constant and return it directly?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        // save/discard buttons
+        propertyActionPanel.getSaveButton().addActionListener(this::onSaveButtonClicked);
+        propertyActionPanel.getDiscardButton().addActionListener(this::onDiscardButtonClicked);
+        // database combox changed
+        databaseComboBox.addItemListener(this::onDatabaseComboBoxChanged);
+    }
+
+    private void onCheckBoxChanged(ItemEvent itemEvent) {
+        if (itemEvent.getStateChange() == ItemEvent.SELECTED || itemEvent.getStateChange() == ItemEvent.DESELECTED) {
+            System.out.println("check box changed...");
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:127]: Remove the `System.out.println`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java (hunk 1 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:100]: the "pattern" parameter is already a pattern and have more than one (two) types currently. more than one constants seems not a good a idea.
these lines means set default values for hostname, database, and username.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzureHideableTitledSeparator.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    public void expand() {
+        this.update(true);
+    }
+
+    public void collapse() {
+        this.update(false);
+    }
+
+    public void addContentComponent(JComponent component) {
+        contentComponentList.add(component);
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzureHideableTitledSeparator.java:58]: should call `update` to make state and view consistent

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    @Setter
+    private Subscription subscription;
+    private Server server;
+
+    public DatabaseComboBox() {
+        super(false);
+    }
+
+    public void setServer(Server server) {
+        if (Objects.equals(server, this.server)) {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java:48]: make sure `Server` overrides `equals`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzureHideableTitledSeparator.java (hunk 3 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/AzureHideableTitledSeparator.java:58]: adding "component.setVisible(isExpanded)"  is enough to sync.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    private ConnectionStringsOutputPanel connectionStringsJDBC;
+    private ConnectionStringsOutputPanel connectionStringsSpring;
+    private JPanel rootPanel;
+    private JPanel contextPanel;
+    private JScrollPane scrollPane;
+    private MySQLPropertyActionPanel propertyActionPanel;
+    private DatabaseComboBox databaseComboBox;
+    private JLabel databaseLabel;
+    public static final String MYSQL_OUTPUT_TEXT_PATTERN_SPRING =
+            "spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver" + StringUtils.LF +
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:69]: what is the difference from 
https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/SystemUtils.html#LINE_SEPARATOR

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/Node.java [hunk 6]
```diff
@@ -204,6 +207,12 @@ public void setIconPath(String iconPath) {
         propertyChangeSupport.firePropertyChange("iconPath", oldValue, iconPath);
     }
 
+    /*public void setAzureIcon(AzureIcon azureIcon) {
```
Comment (Flanker32) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/Node.java:210]: Please remove the comment part

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java [hunk 7]
```diff
@@ -50,7 +48,7 @@
     private ServerState serverState;
 
     public MySQLNode(AzureRefreshableNode parent, String subscriptionId, Server mysqlServer) {
-        super(mysqlServer.id(), mysqlServer.name(), parent, ICON_FILE, true);
+        super(mysqlServer.id(), mysqlServer.name(), parent, null, true);
```
Comment (Flanker32) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:51]: Remove the icon in my sql node?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    protected List<? extends DatabaseInner> loadItems() throws Exception {
+        if (subscription == null) {
+            return new ArrayList<>();
+        }
+        if (server == null || !ServerState.READY.equals(server.userVisibleState())) {
+            return new ArrayList<>();
+        }
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java:74]: may be merged with the upper if block

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java (hunk 7 above)
Comment (shenqianjin) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:51]: Yes. will set a AzureIcon for node on priority.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+        connectionStringsJDBC.getOutputTextArea().setText(getConnectionString(MYSQL_OUTPUT_TEXT_PATTERN_JDBC, null, null, null));
+        connectionStringsSpring.getTitleLabel().setText("Spring");
+        connectionStringsSpring.getOutputTextArea().setText(getConnectionString(MYSQL_OUTPUT_TEXT_PATTERN_SPRING, null, null, null));
+        init();
+        initListeners();
+    }
+
+    private String getConnectionString(final String pattern, final String hostname, final String database, final String username) {
+        if (StringUtils.isBlank(hostname) || StringUtils.isBlank(database) || StringUtils.isBlank(username)) {
+            return String.format(pattern, "{your_hostname}", "{your_database}", "{your_username}");
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:101]: placeholder will be used for `database`/`username` even if `hostname` is blank, but `database`/`username` is not.
is this what you expect?

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:101]: yes. as expected, all the 3 items should be same to be blank or not blank.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    private ConnectionSecurityPanel connectionSecurity;
+    private AzureHideableTitledSeparator connectionStringsSeparator;
+    private ConnectionStringsOutputPanel connectionStringsJDBC;
+    private ConnectionStringsOutputPanel connectionStringsSpring;
+    private JPanel rootPanel;
+    private JPanel contextPanel;
+    private JScrollPane scrollPane;
+    private MySQLPropertyActionPanel propertyActionPanel;
+    private DatabaseComboBox databaseComboBox;
+    private JLabel databaseLabel;
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:68]: I wonder when these components are initialized. do you know?

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:68]: ... they are initialized during constructing.
If you are interested in more detail about GUI, please directly check some GUI references such as https://netbeans.org/kb/docs/java/gui-functionality.html.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+        try {
+            Utils.copyToSystemClipboard(MySQLPropertyView.this.connectionStringsJDBC.getOutputTextArea().getText());
+        } catch (Exception exception) {
+        }
+    }
+
+    private void onSpringCopyButtonClicked(ActionEvent e) {
+        try {
+            Utils.copyToSystemClipboard(MySQLPropertyView.this.connectionStringsSpring.getOutputTextArea().getText());
+        } catch (Exception exception) {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/mysql/MySQLPropertyView.java:144]: IMO. we should tell user if fails to copy into clipboard.

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/AzureAllIcons.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+        public static final AzureIcon OPEN_IN_PORTAL = loadAzureIcon(ICON_BASE_DIR + "mysql/OpenInPortal.svg", ICON_BASE_DIR + "mysql/OpenInPortal_dark.svg");
+
+        public static final AzureIcon SHOW_PROPERTIES = loadAzureIcon(ICON_BASE_DIR + "mysql/ShowProperties.svg",
+                ICON_BASE_DIR + "mysql/ShowProperties_dark.svg");
+    }
+
+    public static final class MySQL {
+
+        private static final String MYSQL_BASE_DIR = ICON_BASE_DIR + "mysql/";
```
Comment (wangmingliang-ms) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/AzureAllIcons.java:49]: seems never used.

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/AzureAllIcons.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+package com.microsoft.tooling.msservices.serviceexplorer;
+
+public class AzureAllIcons {
+
+    private static final String ICON_BASE_DIR = "/icons/";
+
+    public static final class Common {
+
+        public static final AzureIcon CREATE = loadAzureIcon(ICON_BASE_DIR + "mysql/Create.svg", ICON_BASE_DIR + "mysql/Create_dark.svg");
```
Comment (wangmingliang-ms) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/AzureAllIcons.java:31]: why not make a method
```
loadAzureIcon(light, dark){
  loadAzureIconFromPath(ICON_BASE_DIR+light, ICON_BASE_DIR+dark);
}
```

**Review by wangmingliang-ms — APPROVED** [review]
commented.

**Pull request merged**